src/module/OVM_GasPriceOracle.sol

```diff
@@ -0,0 +1,160 @@
+  uint256 public scalar;
+  // Number of decimals of the scalar
+  uint256 public decimals;
+
+  /**
+   * @dev Constructor is left empty and init is used instead
+   */
+  constructor() {}
+
+  /**
+   * @notice Used internally to initialize the contract instead of through a constructor
+   * @dev This function is called by the deployer/factory when creating a contract
+   * @param initPayload abi encoded payload to use for contract initilaization
+   */
+  function init(bytes memory initPayload) external override returns (bytes4) {
+    require(!_isInitialized(), "HOLOGRAPH: already initialized");
+    (uint256 _gasPrice, uint256 _l1BaseFee, uint256 _overhead, uint256 _scalar, uint256 _decimals) = abi.decode(
+      initPayload,
+      (uint256, uint256, uint256, uint256, uint256)
+    );
+    gasPrice = _gasPrice;
+    l1BaseFee = _l1BaseFee;
+    overhead = _overhead;
+    decimals = _decimals;
+    scalar = _scalar;
+    assembly {
+      sstore(_adminSlot, origin())
+    }
+    _setInitialized();
+    return InitializableInterface.init.selector;
+  }
+
+  event GasPriceUpdated(uint256);
+  event L1BaseFeeUpdated(uint256);
+  event OverheadUpdated(uint256);
+  event ScalarUpdated(uint256);
+  event DecimalsUpdated(uint256);
+
+  /**
+   * Allows the owner to modify the l2 gas price.
+   * @param _gasPrice New l2 gas price.
+   */
+  function setGasPrice(uint256 _gasPrice) external onlyAdmin {
+    gasPrice = _gasPrice;
+    emit GasPriceUpdated(_gasPrice);
+  }
+
+  /**
+   * Allows the owner to modify the l1 base fee.
+   * @param _baseFee New l1 base fee
+   */
+  function setL1BaseFee(uint256 _baseFee) external onlyAdmin {
+    l1BaseFee = _baseFee;
+    emit L1BaseFeeUpdated(_baseFee);
+  }
+
+  /**
+   * Allows the owner to modify the overhead.
+   * @param _overhead New overhead
+   */
+  function setOverhead(uint256 _overhead) external onlyAdmin {
+    overhead = _overhead;
+    emit OverheadUpdated(_overhead);
+  }
+
+  /**
+   * Allows the owner to modify the scalar.
+   * @param _scalar New scalar
+   */
+  function setScalar(uint256 _scalar) external onlyAdmin {
+    scalar = _scalar;
+    emit ScalarUpdated(_scalar);
+  }
+
+  /**
+   * Allows the owner to modify the decimals.
+   * @param _decimals New decimals
+   */
+  function setDecimals(uint256 _decimals) external onlyAdmin {
+    decimals = _decimals;
+    emit DecimalsUpdated(_decimals);
+  }
+
+  /**
+   * Computes the L1 portion of the fee
+   * based on the size of the RLP encoded tx
+   * and the current l1BaseFee
+   * @param _data Unsigned RLP encoded tx, 6 elements
+   * @return L1 fee that should be paid for the tx
+   */
+  function getL1Fee(bytes memory _data) external view returns (uint256) {
+    uint256 l1GasUsed = getL1GasUsed(_data);
+    uint256 l1Fee = l1GasUsed * l1BaseFee;
+    uint256 divisor = 10 ** decimals;
+    uint256 unscaled = l1Fee * scalar;
+    uint256 scaled = unscaled / divisor;
+    return scaled;
+  }
+
+  /**
```

Comment:
This comment is super helpful

Reply:
Copy pasta from original Optimism Gas Price Oracle contract. Kept the commentary since it explains the logic very very well.

Reply:
nice